### PR TITLE
Feat: keccak256

### PR DESCRIPTION
**Pull request by 0xurb**
Quick resolve for `KECCAK256` implementation related to https://github.com/r55-eth/r55/issues/4

`cargo run` is passed correctly, but not too sure about this implementation

cc: @leonardoalt

### CODE REVIEW

File: eth-riscv-runtime/src/lib.rs [hunk 1]
```diff
@@ -73,6 +73,36 @@ pub fn revert() -> ! {
...
+        asm!(
+            "ecall",
+            in("a0") offset,
+            in("a1") size,
+            lateout("a0") first,
+            lateout("a1") second,
+            lateout("a2") third,
+            lateout("a3") fourth,
+            in("t0") u32::from(Syscall::Keccak256),
+            options(nostack, preserves_flags)
```
Comment (leonardoalt) [inline eth-riscv-runtime/src/lib.rs:92]: what does this do?

**Review by leonardoalt — COMMENTED** [review]
Sick! Please undraft

File: r55/src/exec.rs [hunk 2]
```diff
@@ -256,6 +257,30 @@ fn execute_riscv(
...
+                        let size: u64 = emu.cpu.xregs.read(11);
+
+                        let data_bytes = if size != 0 {
+                            emu.cpu
+                                .bus
+                                .get_dram_slice(offset..(offset + size))
+                                .unwrap()
+                        } else {
+                            &mut []
+                        };
```
Comment (leonardoalt) [inline r55/src/exec.rs:272]: Can be extracted into a function

File: eth-riscv-syscalls/src/lib.rs [hunk 3]
```diff
@@ -50,11 +50,14 @@ macro_rules! syscalls {
...
+// t0: 5, opcode for caller, returns an address
+// t0: 6, opcode for keccak256, a0: offset, a1: size, returns keccak256 hash
 syscalls!(
     (0, Return, "return"),
     (1, SLoad, "sload"),
     (2, SStore, "sstore"),
     (3, Call, "call"),
     (4, Revert, "revert"),
     (5, Caller, "caller"),
+    (6, Keccak256, "keccak256"),
```
Comment (leonardoalt) [inline eth-riscv-syscalls/src/lib.rs:62]: Let's start using EVM opcodes as syscall ids here

File: r55/src/exec.rs [hunk 4]
```diff
@@ -256,6 +257,30 @@ fn execute_riscv(
                         let third_u64 = u64::from_be_bytes(padded_bytes);
                         emu.cpu.xregs.write(12, third_u64);
                     }
+                    6 => {
```
[the remark this answers is not in the document]
Comment (0xurb) [reply r55/src/exec.rs:260]: Yes thanked about that too
Done it on Odyssey's draft

File: eth-riscv-runtime/src/lib.rs (hunk 1 above)
Comment (0xurb) [reply eth-riscv-runtime/src/lib.rs:92]: - `lateout` - 256 bytes of keccak hash goes to 4 registers, it's like `out("a0"), out("a1") ...` etc. but marked "late" to tell that it only writes to the register after all inputs are read.

- `nostack` - tells the compiler this asm block doesn't use the stack.
- `preserves_flags` - means that `asm!` block does not modify the flags register

Saw `nostack` in other RISC-V adapters code, e.g [this](https://docs.rs/crate/riscv/latest/source/src/asm.rs#156), but it is not necessary I think. Don't have any experience with `asm!`.

File: eth-riscv-syscalls/src/lib.rs (hunk 3 above)
Comment (0xurb) [reply eth-riscv-syscalls/src/lib.rs:62]: @leonardoalt let's do on both PRs (for log also) with opcodes and then in a new PR/commit the rest

**Comment by 0xurb** [comment]
@leonardoalt let's merge both syscalls with ids as regarded [opcodes](https://www.evm.codes) and then make a PR with this improvements:

- replace first five syscalls ids with a regarded [opcodes](https://www.evm.codes)
- replace the hardcoded ids(opcodes) on 
https://github.com/r55-eth/r55/blob/587a4850e293dea4c5ebf5710cf73618eba1509f/r55/src/exec.rs#L154-L155 
by the actual syscall names (from `Syscall` enum) with unhandled syscall opcode exception before matching (use `Syscall ::try_from(opcode).unwrap_or_else(||...)` instead of `_ => {}` match arm)
- extract https://github.com/r55-eth/r55/blob/587a4850e293dea4c5ebf5710cf73618eba1509f/r55/src/exec.rs#L160-L167 into a function

File: eth-riscv-runtime/src/lib.rs (hunk 1 above)
Comment (leonardoalt) [reply eth-riscv-runtime/src/lib.rs:92]: yea I feel like the options are not needed here. I think we should remove it if it works without

**Comment by leonardoalt** [comment]
@0xurb ok sounds good, if you could just remove the asm options and see if it works without, we could remove & merge

**Comment by 0xurb** [comment]
@leonardoalt done, `cargo r` works for me

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**